### PR TITLE
prevent double Close'ing of the signalr and websocket connections

### PR DESCRIPTION
**Pull request by buck54321**
Also some basic Refresh tests, but they didn't catch the bug.

### CODE REVIEW

File: exchanges/exchanges_test.go [hunk 1]
```diff
@@ -231,6 +232,31 @@ func TestPoloniexWebsocket(t *testing.T) {
...
+	resp := &http.Response{
+		Body: io.NopCloser(bytes.NewReader(b)),
+	}
+	d.responses = append(d.responses, resp)
+	return resp
+}
+
 type testBittrexConnection struct {
 	xc           *BittrexExchange
 	subscription hubs.ClientMsg
```
Comment (chappjc) [inline exchanges/exchanges_test.go:262]: current linters on my machine have started fussing about the `conn.subscription = subscription` in Send being ineffective/unused, and it really comes back to this field being unused.  Can we just remove this?

File: exchanges/exchanges_test.go [hunk 2]
```diff
@@ -231,6 +232,31 @@ func TestPoloniexWebsocket(t *testing.T) {
...
+	return resp, nil
+}
+
+func (d *tDoer) queue(body interface{}) *http.Response {
+	b, err := json.Marshal(body)
+	if err != nil {
+		panic("tDoer.queue error:" + err.Error())
+	}
+	resp := &http.Response{
+		Body: io.NopCloser(bytes.NewReader(b)),
```
Comment (chappjc) [inline exchanges/exchanges_test.go:254]: oh yeah go 1.15 still needs ioutil

**Review by chappjc — APPROVED** [review]
bittrex connection is now able to reestablish on the main site.   We can look into signalr replacements without the Close usability issue in the future.

**Comment by xaur** [comment]
```
type Doer interface {
	Do(...) (...)
}
```

Deep stuff.

**Pull request merged**